driver-core/src/main/com/mongodb/client/model/Window.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.client.model;
+
+import com.mongodb.annotations.Beta;
+import org.bson.conversions.Bson;
+
+import java.util.List;
+
+/**
+ * A subset of documents within a partition in the {@link Aggregates#setWindowFields(Object, Bson, List) $setWindowFields} pipeline stage
+ * of an aggregation pipeline (see {@code partitionBy} in {@link Aggregates#setWindowFields(Object, Bson, List)}).
+ *
+ * @see Windows
+ * @since 4.3
+ */
+@Beta
+public interface Window extends Bson {
```

Comment:
I'm not clear that this interface buys us anything.  It has no methods, so it just seems like a way to protect a user from themselves in a way that may prove inconvenient.  It's also arguably inconsistent with the rest of the builder API, which doesn't enforce many type constraints (for example, the sortBy field in setWindowFields is of type Bson, not of type SortBy)

Reply:
I think my main question is will it hamper users in the future? When potentially new window operators are added.  

The magic and lack of safety with Bson is that most things can be Bson (Documents etc..) So methods taking Bson are future flexible because of that.

Reply:
### General thoughts

In Java an interface/class is a language concept that allows us to define a subset of the following: a type, a container for data, a container for instructions. From the perspective of a programmer, a type merely restricts how one may use an object. In spite of the potentially negative connotation of the word "restriction", in programming such restrictions help programmers to know in advance, i.e., before running the program, what can be done with an object and in which context the object can be used. Types provide a guide for a programmer, guarding him from making some mistakes, which is also what any API, including a builder API should strive for.

I did notice that our MQL builder API uses `Bson` and `BsonField` as construction blocks, which is no different from an imaginary SQL builder API that uses `String` as a construction block. Not only such API allows combining blocks that are not combinable, but is also not self-evident, requiring to provide a user with a manual explaining what blocks can be used in what context, because there is no way for a user to extract such information from the API itself.

### Addressing specific concerns

#### Flexibility

> The magic and lack of safety with Bson is that most things can be Bson (Documents etc..) So methods taking Bson are future flexible because of that.

This seems a lot like arguments between weakly- and strongly-typed programming languages. The Java driver is already written in a strongly-typed language, it would be odd to destroy types in the driver API and use only `Object`; even stopping to use generics would be odd, despite being less drastic. What is different about the builder API when it comes to using types?

Note that it is not the case that `Window` "has no methods"--it is `Bson` and it has all the methods `Bson` has. The API implementation does not rely on any arcane knowledge when using a `Window`, which means that even if the `Windows` API would lag behind the server functionality, a user would still be able to construct a `BsonDocument` representing a window and trivially wrap into the `Window` type to use anywhere where the API requires a `Window`. If we will find ourselves constantly lagging behind the server, we may even provide an implementation of a `Window` that wraps a `BsonDocument.

Hopefully it can now be seen that having the `Window` type does not reduces the flexibility compared to using `Bson`/`BsonDocument` instead.

#### API Consistency

While it is true that the API for window functions is less weakly-typed that the rest of the builder API, this fact does not result in different usage patterns and, therefore, does not compel a user to adopt a new style. Indeed, if one were to look at the code that uses this API without storing intermediate building blocks in variables, one would not tell the difference:

```java
group("$localityId",
    Accumulators.sum("totalRainfall", "$rainfall"),
    Accumulators.avg("avgTemp", "$temperature"));

setWindowFields("$localityId", ascending("measurementDateTime"),
        WindowedComputations.sum("monthlyRainfall", "$rainfall", timeRange(-1, CURRENT, MONTH)),
        WindowedComputations.avg("monthlyAvgTemp", "$temperature", timeRange(-1, CURRENT, MONTH)));
```

### Conclusion

Given that moving towards strongly-typed API is beneficial as explained in the "General thoughts", and that it does not even reduce flexibility or introduces new usage patterns, as argued above, I see nothing even remotely sufficient to justify removing the new types.

Reply:
I introduced the `of` builder methods. Spent quite a few time trying to write tiny Scala examples with accordance to how our Scala API is meant to be used, but I still cannot be sure (the code works, that much I am sure). If @rozza were to check the Scala examples, it would be great.